Introduce Puppeteer for handling browsers during tests

timvandermeij opened the pull request:
The commit messages contain more information about the individual changes.

Fixes #11628.

timvandermeij commented:
/botio fonttest

timvandermeij commented:
Well, at least one of the bots is happy. The Windows bot was already broken, so perhaps that fixes itself after a gentle kick.

timvandermeij commented:
/botio-linux unittest

timvandermeij commented:
/botio-linux test

timvandermeij commented:
The browser versions are most likely different from those installed on the bots, so I'm expecting some changes in the reference images here.

timvandermeij commented:
I'm pleasantly surprised that this is working immediately on the Linux bot. It looks like I broke the Windows bot even more, but I'll contact Brendan to get that fixed so we can also trigger the tests there to see how that goes.

Snuffleupagus commented:
This looks like a very good improvement, and the amount of code (or lack thereof) that was required to get this to work is a pleasant surprise!

A couple of questions and things I noticed (based on a quick look):
 - > Moreover, this ensures that contributors always run the tests in both Firefox and Chrome, regardless of which browsers they have installed locally.

   This obviously makes a lot of sense as a default, but could we perhaps add a command-line flag (e.g. `--skip-chromium`) for the test tasks such that there's an easy way to skip Chrome; please?
   To save time/resources, I'm always testing only in Firefox locally (via a suitable browser_manifest.json file) and it would be absolutely great to still be able to do that without having to edit the test files. 

 - Looking at the bot logs, the unit- and font-tests still print their output as before, e.g.
    ```
    Run 1296 tests
    All unit tests passed.
    unit tests runtime was 134.1 seconds
    ```
   However, the `browsertest` no longer appears to log it's success/failure output as before.[1]

---
[1] I'd expect something *similar* to e.g. 
```
OHNOES!  Some tests failed!
  different ref/snapshot: 8
  different first/second rendering: 2
Runtime was 989.1 seconds
[11:41:37] Finished '<anonymous>' after 19 min
[11:41:37] Finished 'bottest' after 19 min
```

timvandermeij commented:
Thank you for the feedback!

> could we perhaps add a command-line flag (e.g. --skip-chromium) for the test tasks such that there's an easy way to skip Chrome

Yes, that sounds like a good idea. I have added it to the to do list.

> However, the browsertest no longer appears to log it's success/failure output as before

That's a good point, I totally missed that. I have no idea where it went, but I'll look into it; it's also on the list now.

Snuffleupagus commented:
One more thing, does benchmarking[1] still work as well? I'd expect the answer to be yes, since everything else seem to work nicely, but please test and/or confirm that that's the case :-)

---
[1] Following the instructions at https://github.com/mozilla/pdf.js/wiki/Benchmarking-your-changes

timvandermeij commented:
/botio-linux test

timvandermeij commented:
/botio-linux test

timvandermeij commented:
/botio-linux test

timvandermeij commented:
/botio-linux test

timvandermeij commented:
This is really strange. The reason why the summary logs for the reference tests aren't there is because the test process apparently stops during the reference test run. However, this does not happen locally at all: all tests are run the summary is logged. I have added extensive logging, but it just stops: no closing logic is called whatsoever...

Snuffleupagus commented:
Wait, is this perhaps simply running into the *limited* log size available on the bots!?
Between the "Extracting..." and "TEST-UNEXPECTED-FAIL | ..." lines, the logs are a lot larger than usual.

timvandermeij commented:
Do you know of a log size limit that we configured somewhere? I'll try to get rid of the "Extracting..." lines to see if that's indeed the case.

Snuffleupagus commented:
> Do you know of a log size limit that we configured somewhere?

No idea unfortunately, but in any case that probably requires access to the bots...

> I'll try to get rid of the "Extracting..." lines to see if that's indeed the case.

If I were to guess, I'd say that the sheer volume of reference test failures might be the biggest problem w.r.t. to the log size. That hypothesis should be easy enough to test though, try running `makeref` and then re-run the `test` afterwards :-)
(Obviously you'd then have to re-run `makeref` against e.g. master afterwards, to restore things to a working state.)

timvandermeij commented:
I have implemented the "Extracting..." log suppression, so let's see if we get further now.

/botio-linux test

timvandermeij commented:
@Snuffleupagus I think you're absolutely right. Having removed the "Extracting..." logs made it go much further. I looked into how such a limit would be configured, but I can't find anything about it. Maybe Node.js decides to kill the process once it reaches a certain limit or something. Anyway, thanks for thinking along since I was a bit stuck here!

Next thing I'm going to try is to use exactly the same versions as currently installed on the bots in the hope to get rid of the failures.

timvandermeij commented:
/botio-linux test

timvandermeij commented on test/test.js:
```diff
@@ -784,52 +760,102 @@ function unitTestPostHandler(req, res) {
+    const oldWrite = process.stdout.write;
+    process.stdout.write = (function(write) {
+      return function(string, encoding, fd) {
+        if (string.includes("Extracting...")) {
```
Done in the new commit.

timvandermeij commented on test/test.js:
```diff
@@ -784,52 +760,102 @@ function unitTestPostHandler(req, res) {
+        if (string.includes("Extracting...")) {
+          return;
+        }
+        write.apply(process.stdout, [string, encoding, fd]);
```
Done in the new commit.

timvandermeij commented:
/botio-linux test

timvandermeij commented:
/botio-windows test

brendandahl commented:
Hmm...looks like firefox stays open on the windows bot and I see:
```
Mon Apr 13 2020 17:56:00 GMT+0000 (Greenwich Mean Time): Command queued at #0 ("cmd_test" from @timvandermeij, id: 7620f52af18fa63)
shell.js: internal error
Error: EPERM: operation not permitted, unlink 'C:\pdfjs\botio-files-pdfjs\private\7620f52af18fa63/node_modules/puppeteer/.local-firefox/win64-76/firefox/browser/features/webcompat@mozilla.org.xpi'
    at Object.unlinkSync (fs.js:956:3)
    at _unlinkSync (C:\pdfjs\botio\node_modules\shelljs\shell.js:1398:10)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1145:9)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
    at rmdirSyncRecursive (C:\pdfjs\botio\node_modules\shelljs\shell.js:1135:7)
```

timvandermeij commented:
I have amended the commit to add a flag that Windows seems to require to be able to start the browser properly, as well as exception handling for when the browser cannot be launched so the process continues nicely instead of failing due to an uncaught promise rejection (causing the bots to wait forever).

timvandermeij commented:
/botio-windows test

timvandermeij commented:
Cool, it's working on Windows now as well. I only have to fix the unhandled promise rejection at the end now, but fortunately it's not fatal and the processes stopped nicely. I have updated the list above.

timvandermeij committed:
Identify browsers using the name instead of the path

The other testing code already uses the name of the browser as the
unique identifier, so I don't see a good reason to not use that for
identifying browsers to quit as well. Doing so simplifies the (already
somewhat complex) testing logic and ensures that we can use existing
functionality (such as the `getSession` function) to retrieve sessions.

timvandermeij committed:
Introduce Puppeteer for handling browsers during tests

This commit replaces our own infrastructure for handling browsers during
tests with Puppeteer. Using our own infrastructure for this had a few
downsides:

- It has proven to not always be reliable, especially when closing the
  browser, causing failures on the bots because browsers were still
  running even though they should have been stopped. Puppeteer should do
  a better job with this because it uses the browser's test built-in
  instrumentation tools for this (the devtools protocol) which our code
  didn't. This also means that we don't have to pass
  parameters/preferences to tweak browser behavior anymore.
- It requires the browsers under test to be installed on the system,
  whereas Puppeteer downloads the browsers before the test. This means
  that setup is much easier (no more manual installations and browser
  manifest files) as well as testing with different browser versions
  (since they can be provisioned on demand). Moreover, this ensures that
  contributors always run the tests in both Firefox and Chrome,
  regardless of which browsers they have installed locally.
- It's all code we have to maintain, so Puppeteer abstracts away how the
  browsers start/stop for us so we don't have to keep that code.

By default, Puppeteer only installs one browser during installation,
hence the need for a post-install script to install the second browser.
This requires `cross-env` to make passing the environment variable work
on both Linux and Windows.

timvandermeij commented:
/botio fonttest

timvandermeij committed:
Implement a command line flag to skip Chrome when running tests

To save time or resources during development it can be useful to run
tests only in Firefox. Previously this could be done by editing the
browser manifest file, but since that file is no longer used for
Puppeteer, this command line flag replaces it. For example, executing
`gulp unittest --noChrome` will only run the unit tests in Firefox.

timvandermeij commented:
/botio-windows unittest

Snuffleupagus commented on gulpfile.js:
```diff
@@ -436,6 +436,9 @@ function createTestSource(testsName, bot) {
     if (bot) {
       args.push("--strictVerify");
     }
+    if (process.argv.includes("--noChrome")) {
+      args.push("--noChrome");
+    }
```
As I hoped, when asking for this feature, it didn't require a lot of code; thanks for doing this!

Just one small thing: Can you please add this same block to the `makeRef` function below as well, such that *all* the various test commands are covered?

timvandermeij replied:
Done in the new commit series.

timvandermeij commented:
/botio test

timvandermeij commented:
/botio-windows fonttest

timvandermeij commented:
/botio fonttest

timvandermeij commented:
This should be ready for prime time now. I have updated the PR description with which upstream issues exist, but they are no show stoppers for us because we have workarounds in place. They are simply filed to improve the upstream code and allow us to remove workarounds once they get fixed upstream.

@Snuffleupagus Would you be willing to review this?

Snuffleupagus reviewed (approved):
The various test-commands still seem to work just fine locally on Windows, which I believe was the most important part for me to check here :-)
Thanks for putting in the, clearly not small, amount of work required to make this happen!

The only slight annoyance I found is that the elements containing logs and ref-test images are now a lot smaller than previously, and doesn't seem to auto-expand with the window size, to the point of being *very* difficult to use.
 - The following DOME element is smaller than previously (during test runs): https://github.com/mozilla/pdf.js/blob/master/test/test_slave.html#L31
 - The DOM element containing the ref-test image comparison is a lot smaller (requiring too much scrolling when comparing failures): https://github.com/mozilla/pdf.js/blob/master/test/resources/reftest-analyzer.html#L53

I suppose that this could be Windows specific, but if there's a simple fix for this that'd be great!
(This is the only thing that I'd *really* appreciate being fixed before this PR lands.)

---

I've not reviewed this in too much detail, since I see the value of actually getting this landed without further delays. (If there's further improvements necessary, they can simply be done in follow-ups.)

timvandermeij commented:
I have fixed the too small elements. It turns out by default Puppeteer uses an 800 x 600 viewport. This is configurable and I have now disabled that to make all elements render normally.

/botio makeref
